### DISCUSSION TITLE
Discovered application names are not same between the 'applications list view' and the 'Graph view'

### DISCUSSION THREAD

**Discussion by qiuyuzhou** [kiali/kiali#5443] [Q&A]
The application list:

![image](https://user-images.githubusercontent.com/2310221/188831464-b3edb770-0f45-4530-bc9d-3b58bd19cf66.png)

The graph view:

![image](https://user-images.githubusercontent.com/2310221/188831821-81af37cc-0ee5-4feb-b143-4a9e6d3530c0.png)

```
{
  "backendConfigs": {
    "authenticationConfig": {
      "strategy": "anonymous"
    },
    "computedServerConfig": {
      "clusters": {
        "prod": {
          "apiEndpoint": "https://192.168.1.117:6443",
          "isKialiHome": false,
          "isGatewayToNamespace": false,
          "kialiInstances": [
            {
              "serviceName": "kiali",
              "namespace": "istio-system",
              "operatorResource": "",
              "url": "",
              "version": "v1.55.0"
            }
          ],
          "name": "prod",
          "network": "",
          "secretName": "istio-remote-secret-prod"
        },
        "prod-22": {
          "apiEndpoint": "https://172.16.0.1:443",
          "isKialiHome": true,
          "isGatewayToNamespace": false,
          "kialiInstances": [
            {
              "serviceName": "kiali",
              "namespace": "istio-system",
              "operatorResource": "",
              "url": "",
              "version": "v1.55.0"
            }
          ],
          "name": "prod-22",
          "network": "hangzhou1",
          "secretName": ""
        }
      },
      "durations": {
        "60": "1m",
        "120": "2m",
        "300": "5m",
        "600": "10m",
        "1800": "30m",
        "3600": "1h",
        "10800": "3h",
        "21600": "6h",
        "43200": "12h",
        "86400": "1d",
        "604800": "7d"
      },
      "healthConfig": {
        "rate": [
          {
            "tolerance": [
              {
                "code": "/5\\d\\d/",
                "degraded": 0,
                "failure": 10,
                "protocol": "/http/",
                "direction": "/.*/"
              },
              {
                "code": "/4\\d\\d/",
                "degraded": 10,
                "failure": 20,
                "protocol": "/http/",
                "direction": "/.*/"
              },
              {
                "code": "/^[1-9]$|^1[0-6]$/",
                "degraded": 0,
                "failure": 10,
                "protocol": "/grpc/",
                "direction": "/.*/"
              },
              {
                "code": "/^-$/",
                "degraded": 0,
                "failure": 10,
                "protocol": "/http|grpc/",
                "direction": "/.*/"
              }
            ],
            "namespace": "/.*/",
            "name": "/.*/",
            "kind": "/.*/"
          }
        ]
      },
      "deployment": {},
      "installationTag": "Kiali Console",
      "istioAnnotations": {
        "istioInjectionAnnotation": "sidecar.istio.io/inject"
      },
      "istioCanaryRevision": {},
      "istioIdentityDomain": "svc.cluster.local",
      "istioNamespace": "istio-system",
      "istioLabels": {
        "appLabelName": "app",
        "injectionLabelName": "istio-injection",
        "injectionLabelRev": "istio.io/rev",
        "versionLabelName": "version"
      },
      "kialiFeatureFlags": {
        "certificatesInformationIndicators": {
          "enabled": true,
          "secrets": [
            "cacerts",
            "istio-ca-secret"
          ]
        },
        "istioInjectionAction": true,
        "istioUpgradeAction": false,
        "uiDefaults": {
          "graph": {
            "findOptions": [
              {
                "description": "Find: slow edges (> 1s)",
                "expression": "rt > 1000"
              },
              {
                "description": "Find: unhealthy nodes",
                "expression": "! healthy"
              },
              {
                "description": "Find: unknown nodes",
                "expression": "name = unknown"
              },
              {
                "description": "Find: nodes with the 2 top rankings",
                "expression": "rank <= 2"
              }
            ],
            "hideOptions": [
              {
                "description": "Hide: healthy nodes",
                "expression": "healthy"
              },
              {
                "description": "Hide: unknown nodes",
                "expression": "name = unknown"
              },
              {
                "description": "Hide: nodes ranked lower than the 2 top rankings",
                "expression": "rank > 2"
              }
            ],
            "settings": {
              "fontLabel": 13,
              "minFontBadge": 7,
              "minFontLabel": 10
            },
            "traffic": {
              "grpc": "requests",
              "http": "requests",
              "tcp": "sent"
            }
          },
          "metricsPerRefresh": "1m",
          "metricsInbound": {},
          "metricsOutbound": {},
          "refreshInterval": "60s"
        },
        "validations": {
          "ignore": [
            "KIA1201"
          ]
        }
      },
      "prometheus": {
        "globalScrapeInterval": 30,
        "storageTsdbRetention": 604800
      },
      "clusterInfo": {
        "name": "prod-22",
        "network": "hangzhou1"
      },
      "istioStatusEnabled": true,
      "istioConfigMap": "istio"
    }
  },
  "currentURL": "http://localhost:20001/kiali/console/graph/namespaces/?traffic=grpc%2CgrpcRequest%2Chttp%2ChttpRequest%2Ctcp%2CtcpSent&graphType=versionedApp&namespaces=prod&duration=300&refresh=60000&layout=kiali-dagre&namespaceLayout=kiali-dagre",
  "reduxState": {
    "authentication": {
      "message": "",
      "session": {
        "username": null,
        "expiresOn": "2022-09-08T08:30:41.897Z"
      },
      "status": 1,
      "uiExpiresOn": "2022-09-08T08:30:41.897Z"
    },
    "globalState": {
      "loadingCounter": 0,
      "isPageVisible": true,
      "kiosk": "",
      "lastRefreshAt": 1662540050065
    },
    "graph": {
      "edgeMode": "all",
      "graphDefinition": {
        "duration": 300,
        "elements": {
          "nodes": [
            {
              "data": {
                "id": "89ce65e246a9e8cdb012b28aaa728a62",
                "nodeType": "box",
                "cluster": "prod",
                "namespace": "",
                "healthData": null,
                "isBox": "cluster"
              }
            },
            {
              "data": {
                "id": "8d9a23ec3a5e6695e2d4ca758b8b4a6c",
                "nodeType": "box",
                "cluster": "prod-22",
                "namespace": "",
                "healthData": null,
                "isBox": "cluster"
              }
            },
 ... (edited by @jmazzitelli -- I removed the thousands of lines of data here, it is irrelevent to the discussion and just made this discussion text extremely large) ...
```

**Comment by jmazzitelli** [comment]
_I am fairly confident this is not a bug, but just a misunderstanding of how things get labeled in the Kiali graph. Moving to Q&A discussion and removing the bug label._

**Comment by jmazzitelli** [comment]
What I see here:
![image](https://user-images.githubusercontent.com/2029470/188875036-8c48a644-5f8a-40e1-8eaf-ecebea0f79ea.png)
tells me you are looking at the "Versioned App" graph. The nodes inside the main `zeus` box are the versions of the workloads in that app (and they all have version `2.0.8` that I see in your screenshot).

The one thing that may be confusing you is this. In your Kiali config I see:

```      "istioLabels": {
        "appLabelName": "app",
...
        "versionLabelName": "version"
      },
```

All of your resources here have both `app` and `app.kubernetes.io/name` labels, and they have different values from one another (i.e. both of those labels always have different values on the same resource -- for example `app=zeus-worker-temporal` and `app.kubernetes.io/name=zeus`

If I look at the `app` label values, none are `zeus`, and yet I see the application boxes labels with `zeus` in the graph when you would expect them to be things like `zeus-worker-temporal` (since that is what your `app` label value is). Why does the graph look like this?

It is most likely because that is how Istio is storing the telemetry. Remember that you have to explicitly declare which app and version labels Kiali should look at, with the defaults being `app` and `version`. However, Istio doesn't work like that. I think what happens is Istio will store in the Prometheus telemetry the name of the app which is taken FIRST from the `app.kubernetes.io/name` and if you do not have that label it will only then look at `app`.  But you DO have `app.kubernetes.io/name` labels, and they are all `zeus` -- and since the graph gets most of its data directly from Istio's telemetry that it stores in Prometheus, that is why you see `zeus` in the graph.

You can confirm this by looking at Prometheus - look at all your Istio timeseries and look for the "destination_app" (UPDATE: I misspoke here earlier - I said `app` and it should be `destination_app`) timeseries label - what values do you see? I suspect you will see `zeus` for many or all of them.

So, the graph looks the way it does because of the telemetry that Istio stores in Prometheus. You can tell Kiali to use the same labels as Istio by setting those config settings in your Kiali CR (i.e. Tell Kiali to look at the `app.kubernetes.io/name` label, NOT `app`).

The other alternative to fix this confusion is to make your `app` label values the same as the `app.kubernetes.io/name` labels. It is actually confusing to have both labels on the same resource but with different values. Typically, you only use one or the other (not both) on the same resource - but if you do have both labels on the same resource, you will want them to have the same value. Otherwise, the confusion becomes - "what is the name of your app?" - because you are indicating two different app names on the same resource. This is the source of confusion between Istio and Kiali - you are defining different app names on the same resource.

Here are some docs you should read about this issue:
* https://kiali.io/docs/faq/graph/#queryprom
* https://kiali.io/docs/configuration/kialis.kiali.io/#.spec.istio_labels.app_label_name
* https://kiali.io/docs/configuration/kialis.kiali.io/#.spec.istio_labels.version_label_name

**Reply by qiuyuzhou** [reply]
> It is most likely because that is how Istio is storing the telemetry. Remember that you have to explicitly declare which app and version labels Kiali should look at, with the defaults being app and version. However, Istio doesn't work like that. I think what happens is Istio will store in the Prometheus telemetry the name of the app which is taken FIRST from the app.kubernetes.io/name and if you do not have that label it will only then look at app. But you DO have app.kubernetes.io/name labels, and they are all zeus -- and since the graph gets most of its data directly from Istio's telemetry that it stores in Prometheus, that is why you see zeus in the graph.

> You can confirm this by looking at Prometheus - look at all your Istio timeseries and look for the "app" timeseries label - what values do you see? I suspect you will see zeus for many or all of them.

My timeseries from promethes:

```
istio_requests_total{
    app="zeus-api", 
    app_kubernetes_io_component="api", 
    app_kubernetes_io_instance="zeus", 
    app_kubernetes_io_name="zeus", 
    connection_security_policy="mutual_tls", 
    destination_app="zeus-api", 
    destination_canonical_revision="v2.0.9", 
    destination_canonical_service="zeus", 
    destination_cluster="prod-22", 
    destination_principal="spiffe://cluster.local/ns/prod/sa/zeus", 
    destination_service="zeus.prod.svc.cluster.local", 
    destination_service_name="zeus", 
    destination_service_namespace="prod", 
    destination_version="v2.0.9", 
    destination_workload="zeus-api", 
    destination_workload_namespace="prod", 
    grpc_response_status="0", 
    instance="zeus", 
    job="kubernetes-pods", 
    namespace="prod", 
    pod="zeus-api-5f6bbc8cd7-d7snt", 
    pod_template_hash="5f6bbc8cd7", 
    reporter="destination", 
    request_protocol="grpc", 
    response_code="200", 
    response_flags="-", 
    security_istio_io_tlsMode="istio", 
    service_istio_io_canonical_name="zeus", 
    service_istio_io_canonical_revision="v2.0.9", 
    source_app="atlas", 
    source_canonical_revision="v1.0.0", 
    source_canonical_service="atlas", 
    source_cluster="prod-22", 
    source_principal="spiffe://cluster.local/ns/prod/sa/atlas", 
    source_version="v1.0.0", 
    source_workload="atlas", 
    source_workload_namespace="prod", 
    topology_istio_io_network="hangzhou1", 
    version="v2.0.9"
}
```

The `app` label value is `zeus-api` instead of the `zeus`.

**Reply by qiuyuzhou** [reply]
This `istio_requests_total{app="zeus"}` query result is empty.

**Reply by jmazzitelli** [reply]
Please read my and jshaughn's explanations below - I provide my own complete description with examples/screenshots that explain the issue.

As jshaughn says, `Istio telemetry supplies the application name in the canonical_service field` -- so look at your timeseries for confirmation:

```
    destination_canonical_service="zeus", 
```

My explanation below provides recommendations for you.

**Reply by qiuyuzhou** [reply]
my `service.istio.io/canonical-name` value is `zeus`, it should be from the service name, not the label `app.kubernetes.io/name` value of pod.

![image](https://user-images.githubusercontent.com/2310221/189098982-72d45d7d-b435-4e10-b116-b7d6d71364b9.png)

**Reply by jmazzitelli** [reply]
Yes, the SERVICE name... but we are discussing the APP name as it appears in the Kiali graph:

![image](https://user-images.githubusercontent.com/2029470/189099339-af80dfcb-9783-49d2-8475-7f5babd036be.png)

See the green "A" - that "zeus" is NOT the service name - it is the APP name.

Let's go back to the beginning. Can you explain WHAT you think is confusing in the graph you posted in the original post? Because that graph you posted is showing a "Versioned App" graph - so it is showing you APPs.

**Reply by jmazzitelli** [reply]
(for example, switch your graph view to "Workload Graph" and then switch to "Service Graph" and look at the differences you see.)

**Comment by jshaughn** [comment]
The explanation above is correct.  And to be specific, Istio telemetry supplies the application name in the `canonical_service` field, and when that value is determined it prefers the label value for `app.kubernetes.io/name` over the label value for `app`.

The graph names are based on the telemetry.  The list names are based on the values of the label names configured for Kiali.

If you are going to keep the `app.kubernetes.io/name` labels you should configure Kiali to use those labels as well.

**Reply by qiuyuzhou** [reply]
I consider kiali should use the value of `destination_app` as the app box name, when generate the graph.

```
istio_requests_total{
    destination_app="zeus-api", 
    destination_canonical_revision="v2.0.9", 
    destination_canonical_service="zeus", 
    destination_cluster="prod-22", 
    destination_principal="spiffe://cluster.local/ns/prod/sa/zeus", 
    destination_service="zeus.prod.svc.cluster.local", 
    destination_service_name="zeus", 
    destination_service_namespace="prod", 
    destination_version="v2.0.9", 
    destination_workload="zeus-api", 
    destination_workload_namespace="prod"
...
}
```

The key difference of this case here is the `service name` is not equal to the `app name`.

**Reply by qiuyuzhou** [reply]
I think kiali should has a test case that the `service name`, `app name` and `workload name` are different with each other.

**Reply by jmazzitelli** [reply]
As I mention in my long comment, you have conflicting data. You are defining different app names on the same resource (your `app` and your `app.kubernetes.io/name` are different and thus in conflict - you essentially have confused Istio and Kiali because of the inconsistent naming you have associated with your resource).

For example, what is your app name for this resource?
![image](https://user-images.githubusercontent.com/2029470/189098250-3a67c4cc-ba58-4ba0-89b2-4372859e5148.png)
 
You have `app=zeus-api` and `app.kubernetes.io/name=zeus` - so is the app name `zeus` or is it `zeus-api`? Kiali and Istio must guess, and they might guess a value you do not want. You must make them identical if you are to have a consistent app name.

**Reply by qiuyuzhou** [reply]
As the timeseries label value from promethes, the `istio` think the app name is `zeus-api` which is from pod label `app`. I think the istio doesn't care the label `app.kubernetes.io/name`.

Please look at the values of timesereis label `destination_app` and `source_app`.

```
istio_requests_total{
    app="zeus-api", 
    app_kubernetes_io_component="api", 
    app_kubernetes_io_instance="zeus", 
    app_kubernetes_io_name="zeus", 
    connection_security_policy="mutual_tls", 
    destination_app="zeus-api", 
    destination_canonical_revision="v2.0.9", 
    destination_canonical_service="zeus", 
    destination_cluster="prod-22", 
    destination_principal="spiffe://cluster.local/ns/prod/sa/zeus", 
    destination_service="zeus.prod.svc.cluster.local", 
    destination_service_name="zeus", 
    destination_service_namespace="prod", 
    destination_version="v2.0.9", 
    destination_workload="zeus-api", 
    destination_workload_namespace="prod", 
    grpc_response_status="0", 
    instance="zeus", 
    job="kubernetes-pods", 
    namespace="prod", 
    pod="zeus-api-5f6bbc8cd7-d7snt", 
    pod_template_hash="5f6bbc8cd7", 
    reporter="destination", 
    request_protocol="grpc", 
    response_code="200", 
    response_flags="-", 
    security_istio_io_tlsMode="istio", 
    service_istio_io_canonical_name="zeus", 
    service_istio_io_canonical_revision="v2.0.9", 
    source_app="atlas", 
    source_canonical_revision="v1.0.0", 
    source_canonical_service="atlas", 
    source_cluster="prod-22", 
    source_principal="spiffe://cluster.local/ns/prod/sa/atlas", 
    source_version="v1.0.0", 
    source_workload="atlas", 
    source_workload_namespace="prod", 
    topology_istio_io_network="hangzhou1", 
    version="v2.0.9"
}
```

**Reply by qiuyuzhou** [reply]
I have tried to dive the code which generate the graph.

Please give a look on below screenshot. Thanks! The marked two lines.

The `lSourceApp` should be the value of `source_app` instead of `source_canonical_service`.
The `lDestApp` should be the value of `destination_app` instead of `destination_service_name`

`graph/telemetry/istio/appender/aggregate_node.go`:

<img width="838" alt="image" src="https://user-images.githubusercontent.com/2310221/189102706-abb190b0-163c-4350-a91f-d90b71e4e160.png">

**Reply by jmazzitelli** [reply]
No that is incorrect. You are not aware of the changes istio made several releases ago. Kiali is graphing the correct telemetry. In fact, istio folks are likely going to remove those destination_app and source_app time series labels. This is what Jay was saying.. istio is now using the canonical time series labels now. Please read my past comments on what you need to do. It is not true istio is ignoring app.kubernetes.io/name..in fact they use it first.

Trust me when I say the kiali code and istio are in line and consistent with each other.

**Reply by jmazzitelli** [reply]
With that said, I am willing for @jshaughn to tell me I'm wrong and that code has a bug.

**Reply by jmazzitelli** [reply]
I'm going to give another example that shows Istio uses `app.kubernetes.io/name` as an override to the old `app` label. Here is my pod, it has `app` and `app.kubernetes.io/name` labels, but they are different from each other:

```
$ kubectl get pods -n simple-mesh simple-server-84799c499c-d9gtp -o jsonpath={.metadata.labels} | jq
{
  "app": "simple-server",
  "app.kubernetes.io/name": "simple",
  "service.istio.io/canonical-name": "simple",
...
}
```

Notice Istio added that service.istio.io/canonical-name label to my pod (I did not add that - Istio did) - and look what value it gave it? It took that from the `app.kubernetes.io/name` NOT `app`.

Here's what my Prometheus timeseries looks like when I query `istio_requests_total{app="simple-server"}`:

`istio_requests_total{app="simple-server", app_kubernetes_io_name="simple", connection_security_policy="mutual_tls", destination_app="simple-server", destination_canonical_revision="v99", destination_canonical_service="simple", destination_cluster="cluster-default", destination_principal="spiffe://cluster.local/ns/simple-mesh/sa/simple-sa", destination_service="simple-server-svc.simple-mesh.svc.cluster.local", destination_service_name="simple-server-svc", destination_service_namespace="simple-mesh", destination_version="v99", destination_workload="simple-server", destination_workload_namespace="simple-mesh", instance="172.17.0.14:15020", job="kubernetes-pods", namespace="simple-mesh", pod="simple-server-84799c499c-d9gtp", pod_template_hash="84799c499c", reporter="destination", request_protocol="http", response_code="200", response_flags="-", security_istio_io_tlsMode="istio", service_istio_io_canonical_name="simple", service_istio_io_canonical_revision="v99", sidecar_istio_io_inject="true", source_app="simple-client", source_canonical_revision="v99", source_canonical_service="simple", source_cluster="cluster-default", source_principal="spiffe://cluster.local/ns/simple-mesh/sa/simple-sa", source_version="v99", source_workload="simple-client", source_workload_namespace="simple-mesh", topology_istio_io_network="network-default", version="v99"}`

Notice specifically these labels:

```
source_canonical_service="simple"
destination_canonical_service="simple"
```

Those are the canonical service names @jshaughn is talking about. And that's where I say Istio is using `app.kubernetes.io/name` - it isn't being ignored; it is being used for the canonical service name.
 
What is confusing about this is the service name is now different than the app name because those two app labels do not match. So Istio's source of the canonical service name (which is `app.kubernetes.io/name` is different than the source for the app name (which is `app`). This is why I say it is recommended to make sure the two are the same - otherwise you get into this confusing mismatch of names.

Again, I ask -- if your   `app` label is "foo" and your `app.kubernetes.io/name` label is different (say, "bar") then  what is your app name?? You have two of them, and they are different. There is no way to discern what the app name really is due to this conflicting metadata on your resource.

**Comment by jmazzitelli** [comment]
I just ran some tests to explain what is going on. I was going to write up an FAQ on this because this sometimes causes confusion. Instead, I'll just point to this discussion if the question comes up again :) We can add an FAQ if we need something more formal in the docs.

I created a simple mesh that contains one client and one server node (the client just sends requests to the server).

In the first test, I have two labels on both nodes: `app` and `app.kubernetes.io/name` where `app` is set to either `simple-client` or `simple-server` and `app.kubernetes.io/name` is always set to `simple`. Kiali will use its default setting for `app_label_name` which is `"app"` (i.e. Kiali will look at the `app` label when it examines resources directly).This is what Kiali looks like when configured like this:

The Graph (notice the app is named "simple" - this is because Istio is looking at `app.kubernetes.io/name` when storing telemetry in Prometheus, and this is where Kiali gets the app name for the graph): 
![simple-app-graph](https://user-images.githubusercontent.com/2029470/188934578-577057cf-7f93-410e-8cc5-d028073243a3.png)

The App List (because this list does not come from telemetry, Kiali needs to read the resource labels directly to determine the app name, and here Kiali is looking at the `app` label to determine what the application name is):
![simple-app-list](https://user-images.githubusercontent.com/2029470/188935632-eeb4ab95-9114-4e48-b638-a3557d022f7c.png)

OK, so clearly this is misconfigured. I am defining different app names in different labels (so which is it? For my client node, is my app name "simple" or is it "simple-client"? Kiali and Istio are confused).

But let's keep the labels and just tell Kiali to look at the `app.kubernetes.io/name` when it needs to determine the app name. In this case, I changed the value of the  Kiali CR's `spec.istio_labels.app_label_name` to be `"app.kubernetes.io/name"`. This means both Kiali and Istio will look at the same label to determine the app name. Even though I still have two different app names in two different labels, at least Kiali and Istio will consistently look at the same label. When I apply this change to the Kiali CR, the graph remains the same, but here is what the app list in Kiali looks like now:
![simple-app-list-single](https://user-images.githubusercontent.com/2029470/188936545-702227dd-4dd0-45ba-92c7-9890cbea9702.png)

Notice there is a single app (which we know consists of two nodes - the client and server) and the name of the app is "simple" (which Kiali takes from the `app.kubernetes.io/name` label now, just as I configured it to do). So now Kiali is consistent in both the graph and its lists.

Let's now try something else. Let's remove the `app.kubernetes.io/name` label from my nodes, leaving only the `app` labels. So my client node will only have `app=simple-client` and my server will only have `app=simple-server`. At this point, Istio will not see any `app.kubernetes.io/name` labels, so it falls back to looking at the `app` labels when storing telemetry for these nodes. Now go reset the Kiali CR's `spec.istio_labels.app_label_name` setting - put it back to be `"app"`. This means Kiali will look at the `app` label when determining application names when it needs to use labels to do so.

Here is what the graph and the app list will look like when you do this:

![simple-app-graph-nolabel](https://user-images.githubusercontent.com/2029470/188937501-9eace2b2-c41b-41ae-a9eb-b7934fb319c2.png)


![simple-app-list-nolabel](https://user-images.githubusercontent.com/2029470/188937523-356f6a63-2832-4ad1-b57b-eb91b99903cc.png)

Notice the graph is now showing both nodes as different apps (because they are considered different apps - their `app` labels have different values). Same with the app list - you now see two applications in the list because, as I say, they are considered two different apps.

Hopefully, this makes it more clear what is going on.

In short:

1. If you have both `app` and `app.kubernetes.io/name` labels on your resources, it is highly recommended that you make sure their values are identical.
2. Make sure you tell Kiali what labels your Istio is using when storing app names in telemetry by setting the Kiali CR `spec.istio_labels` to match your Istio configuraiton. More specifically, if you have `app.kubernetes.io/name` labels on your resources, make sure you tell Kiali to look at those labels by setting `spec.istio_labels.app_label_name` to `app.kubernetes.io/name`. And the same goes for your version labels - set `spec.istio_labels.version_label_name` to `app.kubernetes.io/version` if you put those labels on your resources as opposed to `version`.

**Comment by jshaughn** [comment]
The explanations we have provided are all correct.  There is no bug here.

**Reply by jmazzitelli** [reply]
> The explanations we have provided are all correct. There is no bug here.

Thanks.

One thing that causes confusion is the name of the resource labels versus the timeseries labels - specifically, it is confusing that Istio creates the timeseries label `...canonical_service` but sets its value to something that typically comes from the resource label `app.kubernetes.io/name` or `app`.

So Istio calls it "service" when it is really "app".  That is part of the confusion. It is probably too late to ask the Istio folks to rename that timeseries label. But suffice it to say, that canonical "service" name you see in the timeseries is really the app name.

**Comment by israel-hdez** [comment]
Hi @qiuyuzhou .

I know it is a little bit confusing how both Istio and Kiali work around labeling and naming. Kubernetes gives a lot of flexibility on how to label/name things. Because of this flexibility, maintainers of both Istio and Kiali had to take some "opinionated" decisions.

On that line, first it should be made clear that an "App" is a logical entity in both Istio and Kiali in the sense that Kubernetes is not aware of it. Both Istio and Kiali rely on labels to determine what is part of an App. The topic about which label to use has some historical decisions that changed the preferred label over the time as both Istio and Kiali evolved, and also as [recommended labels](https://kubernetes.io/docs/concepts/overview/working-with-objects/common-labels/) appeared. Because of the historical decisions, Istio can use a set of different labels to use as the "App name" and the set is "prioritized" (i.e. the most preferred one is taken if available). As both Jay and Mazz have mentioned, if you have both `app` and `app.kubernetes.io/name` in your workloads, then Istio picks `app.kubernetes.io/name`.

Note that, AFAIK, the meaning of the labels are only relevant because of _metrics_ and, perhaps, also traces/spans where the app name (and other metadata) needs to be exposed. Out of that context, Istio mostly relies in label selectors and this is why it works without problems. However, Kiali relies a lot on good labeling and as mentioned above, the `destination_canonical_service` and `source_canonical_service` of the metrics is the place where the App names are being reported (despite the confusing naming). We know there are other labels in the metrics, however those are the result of how Istio has evolved and as Mazz mention, their removal is under consideration.

In any case, the way things work is still somewhat opinionated and we are basically asking you to adhere to these conventions to take advantage of Kiali. As has been repeatedly mentioned in the discussion, please choose between using `app` or `app.kubernetes.io/name`, use one of those labels and label your workloads consistently. If you want to use both, make sure that their values are the same. By default, Kiali will use the `app` label on your resources and will try to match/lookup that value on the `*_canonical_service` of the metrics. So, if you choose to use something else than `app`, make sure you configure Kiali as needed.

If you can't do this, well... the weird behavior of Kiali is expected. At this stage we cannot adapt Kiali to your specific needs. Perhaps, your only option will be to fork Kiali and do the changes to make it behave as you desire. But I don't think this is the way to go.

I hope this helps.
